File: 105.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return build(preorder, 0, preorder.size(), inorder, 0, inorder.size());
    }

    TreeNode* build(vector<int>& preorder, int pl, int pr, vector<int>& inorder, int il, int ir){
        if(pr - pl == 0){
            return nullptr;
        }

        TreeNode *node = new TreeNode(preorder[pl]);

        int mid_idx=0;

        for(int i = il; i < ir; ++i){
            if(inorder[i] == preorder[pl]){
                mid_idx = i;
                break;
            }
        }

        node->left = build(preorder, pl+1, pl+mid_idx, inorder, il, il+mid_idx-1);
        node->right = build(preorder, pl+mid_idx+1, pr, inorder, il+mid_idx+1, ir);
        
        return node;
    }
};

File: 105.py
from typing import List
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if len(preorder) == 0:
            return None
        root = TreeNode(preorder[0])
        
        inorder_idx = inorder.index(preorder[0])

        root.left = self.buildTree(preorder[1:inorder_idx+1], inorder[:inorder_idx])
        root.right = self.buildTree(preorder[inorder_idx+1:], inorder[inorder_idx+1:])
        
        return root